### PR TITLE
Add Mana decay computation

### PR DESCRIPTION
**Pull request by thibault-martinez**
https://github.com/iotaledger/tips-draft/blob/tip39/tips/TIP-0039/tip-0039.md#how-to-use-the-lookup-table

### CODE REVIEW

File: sdk/src/types/block/mana/structure.rs [hunk 1]
```diff
@@ -41,14 +41,88 @@ impl ManaStructure {
...
+
+        // we keep applying the lookup table factors as long as n epochs are left
+        let mut remaining_epochs = n;
+
+        while remaining_epochs > 0 {
+            let mut epochs_to_decay = remaining_epochs;
+
+            if epochs_to_decay > 365 {
+                epochs_to_decay = 365
+            }
```
Comment (DaughterOfMars) [inline sdk/src/types/block/mana/structure.rs:113]:
```suggestion
            let epochs_to_decay = remaining_epochs.max(365);
```

Comment (DaughterOfMars) [reply sdk/src/types/block/mana/structure.rs:113]: or should this be using `decay_factors.len()`?

File: sdk/src/types/block/mana/structure.rs [hunk 2]
```diff
@@ -41,14 +41,88 @@ impl ManaStructure {
...
+
+            let decay_factor = self.decay_factor_at(epochs_to_decay).unwrap();
+
+            // apply the decay using fixed-point arithmetics.
+            (value_hi, value_lo) =
+                multiplication_and_shift(value_hi, value_lo, decay_factor, self.decay_factors().len() as u8);
+        }
+
+        // combine both uint64 variables to get the actual value
+        value_hi << 32 + value_lo
```
Comment (DaughterOfMars) [inline sdk/src/types/block/mana/structure.rs:124]:
```suggestion
        value_hi << 32 | value_lo
```

File: sdk/src/types/block/mana/structure.rs [hunk 3]
```diff
@@ -41,14 +41,88 @@ impl ManaStructure {
...
+    pub fn decay(&self, value: u64, n: u64) -> u64 {
+        /// Returns the upper n bits of a u64 value.
+        fn upper_bits(v: u64, n: usize) -> u64 {
+            v >> n
+        }
+
+        /// Returns the lower n bits of a u64 value.
+        fn lower_bits(mut v: u64, n: usize) -> u64 {
+            v = v << (64 - n);
+            v >> (64 - n)
```
Comment (DaughterOfMars) [inline sdk/src/types/block/mana/structure.rs:62]:
```suggestion
            v & 0xFFFFFFFF
```

**Comment by thibault-martinez** [comment]
Closing in favour of https://github.com/iotaledger/iota-sdk/pull/1376

**Pull request closed without merging**